File: USACO/Platinum/2016-2017/grass.java
import java.util.*;
import java.io.*;

//idea : https://codeforces.com/blog/entry/50889?#comment-348725

//TLE :( but same idea gets AC in c++

public class grass {
	final static int MOD = 1000000007;
	final static int intMax = 10000000;
	final static int off = 100000000;
	final static int intMin = -1000000000;
	final static int[] DX = { 0, 0, -1, 1 };
	final static int[] DY = { -1, 1, 0, 0 };

	static class Reader {
		final private int BUFFER_SIZE = 1 << 16;
		private DataInputStream din;
		private byte[] buffer;
		private int bufferPointer, bytesRead;

		public Reader() {
			din = new DataInputStream(System.in);
			buffer = new byte[BUFFER_SIZE];
			bufferPointer = bytesRead = 0;
		}

		public Reader(String file_name) throws IOException {
			din = new DataInputStream(new FileInputStream(file_name));
			buffer = new byte[BUFFER_SIZE];
			bufferPointer = bytesRead = 0;
		}

		public String readLine() throws IOException {
			byte[] buf = new byte[360]; // line length
			int cnt = 0, c;
			while ((c = read()) != -1) {
				if (c == '\n')
					break;
				buf[cnt++] = (byte) c;
			}
			return new String(buf, 0, cnt);
		}

		public int nextInt() throws IOException {
			int ret = 0;
			byte c = read();
			while (c <= ' ')
				c = read();
			boolean neg = (c == '-');
			if (neg)
				c = read();
			do {
				ret = ret * 10 + c - '0';
			} while ((c = read()) >= '0' && c <= '9');

			if (neg)
				return -ret;
			return ret;
		}

		public long nextLong() throws IOException {
			long ret = 0;
			byte c = read();
			while (c <= ' ')
				c = read();
			boolean neg = (c == '-');
			if (neg)
				c = read();
			do {
				ret = ret * 10 + c - '0';
			} while ((c = read()) >= '0' && c <= '9');
			if (neg)
				return -ret;
			return ret;
		}

		public double nextDouble() throws IOException {
			double ret = 0, div = 1;
			byte c = read();
			while (c <= ' ')
				c = read();
			boolean neg = (c == '-');
			if (neg)
				c = read();

			do {
				ret = ret * 10 + c - '0';
			} while ((c = read()) >= '0' && c <= '9');

			if (c == '.') {
				while ((c = read()) >= '0' && c <= '9') {
					ret += (c - '0') / (div *= 10);
				}
			}

			if (neg)
				return -ret;
			return ret;
		}

		private void fillBuffer() throws IOException {
			bytesRead = din.read(buffer, bufferPointer = 0, BUFFER_SIZE);
			if (bytesRead == -1)
				buffer[0] = -1;
		}

		private byte read() throws IOException {
			if (bufferPointer == bytesRead)
				fillBuffer();
			return buffer[bufferPointer++];
		}

		public void close() throws IOException {
			if (din == null)
				return;
			din.close();
		}
	}
	
	static class Sb_Output implements Closeable, Flushable{
	    StringBuilder sb;
	    OutputStream os;
	    int BUFFER_SIZE;
	    public Sb_Output(String s) throws Exception {
	    	os = new BufferedOutputStream(new FileOutputStream(new File(s)));
	    	BUFFER_SIZE = 1 << 16;
	    	sb = new StringBuilder(BUFFER_SIZE);
	    }
	    public void print(String s) {
	        sb.append(s);
	        if(sb.length() >= (BUFFER_SIZE>>1)) {
	            flush();
	        }
	    }
	    public void println(String s) {
	        print(s);
	        print("\n");
	    }
	    public void flush() {
	        try {
	            os.write(sb.toString().getBytes());
	            sb = new StringBuilder(BUFFER_SIZE);
	        }catch(IOException e) {
	            e.printStackTrace();
	        }
	    }
	    public void close() {
	        flush();
	        try {
	            os.close();
	        }catch(IOException e) {
	            e.printStackTrace();
	        }
	    }
	}
	
	static int max(int a, int b) {
		return a > b ? a : b;
	}
	
	static int min(int a, int b) {
		return a < b ? a : b;
	}
	
	static int N, M, K, Q;
	static ArrayList<edge>[] adj;
	static query[] queries;
	static int[] neworder, order, par, types, orderToActual, loc, weights;
	static int ind = 0;
	static TreeSet<range>[] ranges;
	static fullEdge edges[];
	static int[] parent, rank;

	static void dfs(int v, int p) {
		par[v] = max(0, p);
		if (p == -1)
			order[ind++] = v;
		ArrayList<Integer> children = new ArrayList<Integer>();
		for (edge i : adj[v]) {
			if (i.dest == p)
				continue;
			children.add(i.dest);
		}
		Collections.sort(children, (a, b) -> types[a] - types[b]);
		ArrayList<range> cands = new ArrayList<range>();
		int cur = -1;
		for (Integer i : children) {
			if (cur == -1) {
				cur = types[i];
				cands.add(new range(cur, ind));
			}
			if (types[i] != cur) {
				cands.get(cands.size() - 1).r = ind - 1;
				cands.add(new range(types[i], ind));
				cur = types[i];
			}
			order[ind++] = i;
		}
		if (cands.size() > 0)
			cands.get(cands.size() - 1).r = ind - 1;
		for (range i : cands) {
			ranges[v].add(i);
		}
		for (edge i : adj[v]) {
			if (i.dest == p)
				continue;
			weights[i.dest] = i.weight;
			dfs(i.dest, v);
		}
	}
	

	public static void main(String[] args) throws Exception {
//		Reader in = new Reader();
		Reader in = new Reader("grass.in");
		Sb_Output pw = new Sb_Output("grass.out");
		N = in.nextInt();
		M = in.nextInt();
		K = in.nextInt();
		Q = in.nextInt();
		adj = new ArrayList[N];
		parent = new int[N];
		rank = new int[N];
		queries = new query[Q];
		par = new int[N];
		order = new int[N];
		ranges = new TreeSet[N];
		neworder = new int[N + Q];
		orderToActual = new int[N];
		loc = new int[N];
		weights = new int[N];
		for (int i = 0; i < N; ++i) {
			adj[i] = new ArrayList<edge>();
			ranges[i] = new TreeSet<range>((a, b) -> a.type - b.type);
		}
		edges = new fullEdge[M];
		for (int i = 0; i < M; ++i) {
			int A = in.nextInt() - 1;
			int B = in.nextInt() - 1;
			int L = in.nextInt();
			edges[i] = new fullEdge(A, B, L);
		}
		types = new int[N];
		for (int i = 0; i < N; ++i) {
			types[i] = in.nextInt();
		}
		for (int i = 0; i < Q; ++i) {
			int A = in.nextInt() - 1;
			int B = in.nextInt();
			queries[i] = new query(A, B, i);
		}
		KruskalsAlgorithm();
		LazySegmentTree lst = new LazySegmentTree(N + Q);
		dfs(0, -1);
		for (int i = 0; i < N; ++i) {
			orderToActual[order[i]] = i;
		}
		Arrays.sort(queries, (a, b) -> orderToActual[par[a.node]] == orderToActual[par[b.node]] ? a.type - b.type
				: orderToActual[a.node] - orderToActual[b.node]);
		Arrays.fill(neworder, -1);
		int amtforward = 0;
		two[] stats = new two[N + Q];
		for (int i = 0; i < Q; ++i) {
			if (queries[i].node == 0)
				continue;
			int curnode = queries[i].node;
			int curtype = queries[i].type;
			range dummy = new range(curtype, 0);
			range next = ranges[par[curnode]].ceiling(dummy);
			if (next == null) {
				next = ranges[par[curnode]].floor(dummy);
				neworder[next.r + amtforward + 1] = -2;
				stats[next.r + amtforward + 1] = new two(par[curnode], curtype);
				queries[i].move = next.r + amtforward + 1;
			} else {
				neworder[next.l + amtforward] = -2;
				queries[i].move = next.l + amtforward;
				stats[next.l + amtforward] = new two(par[curnode], curtype);
			}
			++amtforward;
		}
		amtforward = 0;
		// -2 if filled with a query
		for (int i = 0; i < N; ++i) {
			while (neworder[i + amtforward] == -2)
				++amtforward;
			neworder[i + amtforward] = order[i];
		}
		for (int i = 0; i < N + Q; ++i) {
			if (neworder[i] <= 0) {
				lst.update(i, i, off);
			} else {
				stats[i] = new two(par[neworder[i]], types[neworder[i]]);
				loc[neworder[i]] = i;
				lst.update(i, i, weights[neworder[i]]);
			}
		}
		// update ranges
		for (int i = 0; i < N; ++i) {
			ranges[i].clear();
		}
		for (int i = 0; i < N + Q; ++i) {
			if (stats[i] == null)
				continue;
			int curpar = stats[i].par;
			int curtype = stats[i].type;
			int ind = i;
			while (ind < N + Q && stats[ind] != null && stats[ind].par == curpar && stats[ind].type == curtype) {
				++ind;
			}
			ranges[curpar].add(new range(curtype, i, ind - 1));
			i = ind - 1;
		}
		// done updating ranges
		for (int i = 0; i < N; ++i) {
			if (i != 0) {
				if (types[i] == types[par[i]]) {
					lst.update(loc[i], loc[i], intMax);
				}
			}
		}

		Arrays.sort(queries, (a, b) -> a.ind - b.ind);
		for (int i = 0; i < Q; ++i) {
			//p11
			int node = queries[i].node;
			int oldtype = types[node];
			types[node] = queries[i].type;
			int newtype = types[node];
			int move = queries[i].move;
			if (node != 0) {
				lst.update(loc[node], loc[node], 2 * intMax);
				loc[node] = move;
				if (newtype == types[par[node]]) {
					lst.update(move, move, -lst.minQuery(move, move) + intMax + weights[node]);
				}
				else {
					lst.update(move, move, -lst.minQuery(move, move) + weights[node]);
				}
			}
			
			range dummy1 = new range(oldtype, 0); // remove intmax from these
			range remove = ranges[node].ceiling(dummy1);
			if (remove != null && remove.type == oldtype) {
				lst.update(remove.l, remove.r, -intMax);
			}
			range dummy2 = new range(newtype, 0); // add intmax to these
			range add = ranges[node].ceiling(dummy2);
			if (add != null && add.type == newtype) {
				lst.update(add.l, add.r, intMax);
			}
			int ans = lst.minQuery(0, N + Q - 1);
			if (ans < intMax) {
				pw.println("" + ans);
			}
		}
		pw.close();
		in.close();
	}

	static class two {
		int par, type;

		two(int p, int t) {
			par = p;
			type = t;
		}
	}

	static class range {
		int type, l, r;

		range(int c, int a) {
			type = c;
			l = a;
		}

		range(int c, int a, int b) {
			type = c;
			l = a;
			r = b;
		}

		public String toString() {
			return type + " " + l + " " + r;
		}
	}

	static class query {
		int node, type, move, ind;

		query(int n, int c, int i) {
			node = n;
			type = c;
			ind = i;
		}

		public String toString() {
			return node + " " + type + " " + move;
		}
	}

	static class edge {
		int dest, weight;

		edge(int d, int w) {
			dest = d;
			weight = w;
		}
		public String toString() {
			return dest + " " + weight;
		}
	}

	static class fullEdge {
		int src, dest, weight;

		fullEdge(int s, int d, int w) {
			src = s;
			dest = d;
			weight = w;
		}
	}

	static int find(int x) {
		if (parent[x] != x) {
			parent[x] = find(parent[x]);
		}
		return parent[x];
	}

	static void union(int x, int y) {
		int xRoot = find(x), yRoot = find(y);
		if (xRoot == yRoot)
			return;
		if (rank[xRoot] < rank[yRoot])
			parent[xRoot] = yRoot;
		else if (rank[yRoot] < rank[xRoot])
			parent[yRoot] = xRoot;
		else {
			parent[yRoot] = xRoot;
			rank[xRoot] = rank[xRoot] + 1;
		}
	}

	static void KruskalsAlgorithm() {
		Arrays.sort(edges, (a, b) -> a.weight - b.weight);
		for (int i = 0; i < N; ++i) {
			parent[i] = i;
		}
		int numEdges = 0;
		int edgeIndex = 0;
		while (numEdges < N - 1) {
			fullEdge next = edges[edgeIndex];
			int x = find(next.src);
			int y = find(next.dest);
			int sc = next.src;
			int dst = next.dest;
			int wt = next.weight;
			if (x != y) {
				adj[sc].add(new edge(dst, wt));
				adj[dst].add(new edge(sc, wt));
				union(x, y);
				++numEdges;
			}
			++edgeIndex;
		}
	}
	static class LazySegmentTree {
		private int[] min;
		private int[] lazy;
		private int size;

		public LazySegmentTree(int size) {
			this.size = size;
			min = new int[4 * size];
			lazy = new int[4 * size];
		}

		public void update(int l, int r, int inc) {
			update(1, 0, size - 1, l, r, inc);
		}

		private void pushDown(int index, int l, int r) {
			min[index] += lazy[index];
			if (l != r) {
				lazy[2 * index] += lazy[index];
				lazy[2 * index + 1] += lazy[index];
			}
			lazy[index] = 0;
		}

		private void pullUp(int index, int l, int r) {
			int m = (l + r) / 2;
			min[index] = min(evaluateMin(2 * index, l, m), evaluateMin(2 * index + 1, m + 1, r));
		}

		private int evaluateMin(int index, int l, int r) {
			return min[index] + lazy[index];
		}
		
		private void update(int index, int l, int r, int left, int right, int inc) {
			if (r < left || l > right)
				return;
			if (l >= left && r <= right) {
				lazy[index] += inc;
				return;
			}
			pushDown(index, l, r);
			int m = (l + r) / 2;
			update(2 * index, l, m, left, right, inc);
			update(2 * index + 1, m + 1, r, left, right, inc);
			pullUp(index, l, r);
		}

		public int minQuery(int l, int r) {
			return minQuery(1, 0, size - 1, l, r);
		}

		private int minQuery(int index, int l, int r, int left, int right) {
			if (r < left || l > right)
				return Integer.MAX_VALUE;
			if (l >= left && r <= right) {
				return evaluateMin(index, l, r);
			}
			pushDown(index, l, r);
			int m = (l + r) / 2;
			int ret = Integer.MAX_VALUE;
			ret = min(ret, minQuery(2 * index, l, m, left, right));
			ret = min(ret, minQuery(2 * index + 1, m + 1, r, left, right));
			pullUp(index, l, r);
			return ret;
		}
	}
	
}

File: USACO/Platinum/2016-2017/grass.cpp
//this idea is used: https://codeforces.com/blog/entry/50889?#comment-348725
// AC !!!
#include <bits/stdc++.h>
using namespace std;
 
using ll = long long;
using pii = pair<int, int>;
 
#define pb push_back
#define mp make_pair
#define eb emplace_back
#define all(x) (x).begin(),(x).end()
#define x first
#define y second
 
const int MOD = 1e9 + 7;
const int mxN = 2e5 + 5;
const int mxQ = 2e5 + 5;
const int mxM = 2e5 + 5;
const int intMax = 1e7;
const int off = 1e8;

void setIO(string s) {
    ios_base::sync_with_stdio(0);
    cin.tie(0);
    freopen((s + ".in").c_str(), "r", stdin);
    freopen((s + ".out").c_str(), "w", stdout);
}

void setIO2() {
    ios_base::sync_with_stdio(0);
    cin.tie(0);
//    freopen((s + ".in").c_str(), "r", stdin);
//    freopen((s + ".out").c_str(), "w", stdout);
}

struct two{
    int par, type;
    two(){
        par = -1; type = -1;
    }
    two(int a, int b){
        par = a; type = b;
    }
};

struct range{
    int type, l, r;
    range(){}
    range(int a, int b){
        type = a; l = b;
    }
    range(int a, int b, int c){
        type = a; l = b; r = c;
    }
    bool operator<(const range o)const{
        return type < o.type;
    }
};

struct query{
    int node, type, move, ind;
    query(){}
    query(int a, int b, int d){
        node = a; type = b; ind = d;
    }
    query(int a, int b, int c, int d){
        node = a; type = b; move = c; ind = d;
    }
};

struct edge{
    int dest, weight;
    edge(){}
    edge(int a, int b){
        dest = a; weight = b;
    }
};

struct fullEdge{
    int src, dest, weight;
    fullEdge(){}
    fullEdge(int a, int b, int c){
        src = a; dest = b; weight = c;
    }
    bool operator<(const fullEdge o) const{
        return weight < o.weight;
    }
};
struct LazySegmentTree {
    int mn[4 * (mxN + mxQ)];
    int lazy[4 * (mxN + mxQ)];
    int size;
    LazySegmentTree(int sz) {
        size = sz;
    }
    void update(int l, int r, int inc) {
        update(1, 0, size - 1, l, r, inc);
    }
    void pushDown(int index, int l, int r) {
        mn[index] += lazy[index];
        if (l != r) {
            lazy[2 * index] += lazy[index];
            lazy[2 * index + 1] += lazy[index];
        }
        lazy[index] = 0;
    }

    int evaluateMin(int index, int l, int r) {
        return mn[index] + lazy[index];
    }

    void pullUp(int index, int l, int r) {
        int m = (l + r) / 2;
        mn[index] = min(evaluateMin(2 * index, l, m), evaluateMin(2 * index + 1, m + 1, r));
    }
    
    void update(int index, int l, int r, int left, int right, int inc) {
        if (r < left || l > right)
            return;
        if (l >= left && r <= right) {
            lazy[index] += inc;
            return;
        }
        pushDown(index, l, r);
        int m = (l + r) / 2;
        update(2 * index, l, m, left, right, inc);
        update(2 * index + 1, m + 1, r, left, right, inc);
        pullUp(index, l, r);
    }
    int minQuery(int index, int l, int r, int left, int right) {
        if (r < left || l > right)
            return 1e9;
        if (l >= left && r <= right) {
            return evaluateMin(index, l, r);
        }
        pushDown(index, l, r);
        int m = (l + r) / 2;
        int ret = 1e9;
        ret = min(ret, minQuery(2 * index, l, m, left, right));
        ret = min(ret, minQuery(2 * index + 1, m + 1, r, left, right));
        pullUp(index, l, r);
        return ret;
    }
    int minQuery(int l, int r) {
        return minQuery(1, 0, size - 1, l, r);
    }
};

int N, M, K, Q;
vector<edge> adj[mxN];
query queries[mxQ];
int neworder[mxN + mxQ], order[mxN], par[mxN], types[mxN], orderToActual[mxN], loc[mxN], weights[mxN], parent[mxN], rnk[mxN];
int ind = 0;
set<range> ranges[mxN];
fullEdge edges[mxM];
two stats[mxN + mxQ];

bool dfsComp(int a, int b){
    return types[a] < types[b];
}

bool queryComp1(query a, query b){
    return orderToActual[par[a.node]] == orderToActual[par[b.node]] ? a.type < b.type : orderToActual[a.node] < orderToActual[b.node];
}

bool queryComp2(query a, query b){
    return a.ind < b.ind;
}

int find(int x) {
    if (parent[x] != x) {
        parent[x] = find(parent[x]);
    }
    return parent[x];
}

void merge(int x, int y) {
    int xRoot = find(x), yRoot = find(y);
    if (xRoot == yRoot)
        return;
    if (rnk[xRoot] < rnk[yRoot])
        parent[xRoot] = yRoot;
    else if (rnk[yRoot] < rnk[xRoot])
        parent[yRoot] = xRoot;
    else {
        parent[yRoot] = xRoot;
        rnk[xRoot] = rnk[xRoot] + 1;
    }
}

void KruskalsAlgorithm() {
    sort(edges, edges + M);
    for (int i = 0; i < N; ++i) {
        parent[i] = i;
    }
    int numEdges = 0;
    int edgeIndex = 0;
    while (numEdges < N - 1) {
        fullEdge next = edges[edgeIndex];
        int x = find(next.src);
        int y = find(next.dest);
        int sc = next.src;
        int dst = next.dest;
        int wt = next.weight;
        if (x != y) {
            adj[sc].eb((edge){dst, wt});
            adj[dst].eb((edge){sc, wt});
            merge(x, y);
            ++numEdges;
        }
        ++edgeIndex;
    }
}

void dfs(int v, int p){
    par[v] = max(0, p);
    if(p == -1){
        order[ind++] = v;
    }
    vector<int> children;
    for(edge i : adj[v]){
        if(i.dest == p) continue;
        children.eb(i.dest);
    }
    sort(all(children), dfsComp);
    vector<range> cands;
    int cur = -1;
    for(int i : children){
        if(cur == -1){
            cur = types[i];
            cands.eb((range){cur, ind});
        }
        if(types[i] != cur){
            cands[cands.size() - 1].r = ind - 1;
            cands.eb((range){types[i], ind});
            cur = types[i];
        }
        order[ind++] = i;
    }
    if(cands.size()){
        cands[cands.size() - 1].r = ind - 1;
    }
    for(range i : cands){
        ranges[v].insert(i);
    }
    for(edge i : adj[v]){
        if(i.dest == p) continue;
        weights[i.dest] = i.weight;
        dfs(i.dest, v);
    }
}

int main(){
    setIO("grass");
//    setIO2();
    cin >> N >> M >> K >> Q;
    for(int i = 0; i < M; ++i){
        int a, b, l; cin >> a >> b >> l;
        --a; --b;
        edges[i] = (fullEdge){a, b, l};
    }
    for(int i = 0; i < N; ++i){
        cin >> types[i];
    }
    for(int i = 0; i < Q; ++i){
        int a, b; cin >> a >> b;
        --a;
        queries[i] = (query){a, b, i};
    }
    KruskalsAlgorithm();
    LazySegmentTree lst(N + Q);
    dfs(0, -1);
    for(int i = 0; i < N; ++i){
        orderToActual[order[i]] = i;
    }
    sort(queries, queries + Q, queryComp1);
    fill(neworder, neworder + N + Q, -1);
    int amtforward = 0;
    for(int i = 0; i < Q; ++i){
        if(queries[i].node == 0)
            continue;
        int curnode = queries[i].node;
        int curtype = queries[i].type;
        range dummy = (range){curtype, 0};
        auto it = ranges[par[curnode]].lower_bound(dummy);
        range next = *it;
        if(it == ranges[par[curnode]].end()) {
            --it;
            next = *it;
            neworder[next.r + amtforward + 1] = -2;
            stats[next.r + amtforward + 1] = (two){par[curnode], curtype};
            queries[i].move = next.r + amtforward + 1;
        } else {
            neworder[next.l + amtforward] = -2;
            queries[i].move = next.l + amtforward;
            stats[next.l + amtforward] = (two){par[curnode], curtype};
        }
        ++amtforward;
    }
    amtforward = 0;
    for (int i = 0; i < N; ++i) {
        while (neworder[i + amtforward] == -2)
            ++amtforward;
        neworder[i + amtforward] = order[i];
    }
    for (int i = 0; i < N + Q; ++i) {
        if (neworder[i] <= 0) {
            lst.update(i, i, off);
        } else {
            stats[i] = (two){par[neworder[i]], types[neworder[i]]};
            loc[neworder[i]] = i;
            lst.update(i, i, weights[neworder[i]]);
        }
    }
    for (int i = 0; i < N; ++i) {
        ranges[i].clear();
    }
    for (int i = 0; i < N + Q; ++i) {
        if (stats[i].type == -1)
            continue;
        int curpar = stats[i].par;
        int curtype = stats[i].type;
        int ind = i;
        while (ind < N + Q && stats[ind].type != -1 && stats[ind].par == curpar && stats[ind].type == curtype) {
            ++ind;
        }
        ranges[curpar].insert((range){curtype, i, ind - 1});
        i = ind - 1;
    }
    for (int i = 0; i < N; ++i) {
        if (i != 0) {
            if (types[i] == types[par[i]]) {
                lst.update(loc[i], loc[i], intMax);
            }
        }
    }
    sort(queries, queries + Q, queryComp2);
    for(int i = 0; i < Q; ++i){
        int node = queries[i].node;
        int oldtype = types[node];
        types[node] = queries[i].type;
        int newtype = types[node];
        int move = queries[i].move;
        if (node != 0) {
            lst.update(loc[node], loc[node], 2 * intMax);
            loc[node] = move;
            if (newtype == types[par[node]]) {
                lst.update(move, move, -lst.minQuery(move,move) + intMax + weights[node]);
            }
            else {
                lst.update(move, move, -lst.minQuery(move,move) + weights[node]);
            }
        }
        range dummy1 = (range){oldtype, 0}; // remove intmax from these
        auto it = ranges[node].lower_bound(dummy1);
        range remove = *it;
        if (it != ranges[node].end() && remove.type == oldtype) {
            lst.update(remove.l, remove.r, -intMax);
        }
        range dummy2 = (range){newtype, 0}; // add intmax to these
        auto it2 = ranges[node].lower_bound(dummy2);
        range add = *it2;
        if (it2 != ranges[node].end() && add.type == newtype) {
            lst.update(add.l, add.r, intMax);
        }
        int ans = lst.minQuery(0, N + Q - 1);
        if (ans < intMax) {
            cout << ans << "\n";
        }
    }
}